fms/utils/serialization.py

```diff
@@ -219,7 +219,7 @@ def load_state_dict(
     else:
         with torch.no_grad():
             checkpoint_sds = [
-                torch.load(str(ckpt_path), mmap=True) for ckpt_path in checkpoints
+                torch.load(str(ckpt_path), map_location=initial_device, mmap=True) for ckpt_path in checkpoints
```

Comment:
hmmm... I wonder if this would introduce any risk of OOM issues.

We create an instance of a model on-device (with empty uninitialized tensors but that take up device memory) and then copy the state_dict into it. If the state_dict is initially loaded onto the device, then we'll have two copies of every tensor in GPU memory, whereas if it's initially loaded onto CPU we minimize peak GPU memory usage.

Reply:
If we do this, we should also redo our model initialization to be on meta device when loading a state dict to avoid these OOM issues (and then assign the weights instead of using `copy_` as we currently do). But given how we initialize things like RoPE caches that might need some extra careful tinkering. Maybe have some kind of `__device_init__` function in our model that only executes when we move away from the meta device? I know Pytorch's FSDP wrapper has something like that for the same reason.

It was actually on my TODO list to try something like this (hence why I have already given it some thought), but given my current workload you can take a stab at it @puneeshkhanna if you want.

Reply:
Yes even 70B also gave OOM. Let me think a bit more about this and my workload too. Btw right now isn't torch.load loading directly on device only and it seems to be taking CUDA by default.